src/exporters/vhi.py

```diff
@@ -0,0 +1,220 @@
+
+# TODO: separate into general ftp exporter class?
+# class FTPExporter():
+#     """"""
+
+# ------------------------------------------------------------------------------
+# Parallel functions
+# TODO: do these functions generalise? If so they can go into a utils.py
+# ------------------------------------------------------------------------------
+
+
+def _parse_time_from_filename(filename) -> Tuple:
+    # regex pattern (4 digits after '.P')
+    year_pattern = re.compile(r'.P\d{4}')
+    # extract the week_number
+    week_num = year_pattern.split(filename)[-1].split('.')[0]
+    # extract the year from the filename
+    year = year_pattern.findall(filename)[0].split('.P')[-1]
+
+    return year, week_num
+
+
+def make_filename(raw_folder: Path, raw_filename: str, dataset: str = 'vhi',) -> Path:
+    # check that the string is a legitimate name
+    assert len(raw_filename.split('/')) == 1, f" \
+        filename cannot have subdirectories in it '/'. Must be the root \
+        filename. Currently: {raw_filename}\
+        "
+
+    # make the dataset folder ('VHI')
+    dataset_folder = raw_folder / dataset
+    if not dataset_folder.exists():
+        dataset_folder.mkdir()
+
+    # make the year folder
+    year = _parse_time_from_filename(raw_filename)[0]
+    assert isinstance(year, str), f"year must be a string! currently: < {year}, {type(year)} >"
+    year_folder = dataset_folder / year
+    if not year_folder.exists():
+        year_folder.mkdir()
+
+    # make the filename e.g. 'raw/vhi/1981/VHP.G04.C07.NC.P1981035.VH.nc'
+    filename = year_folder / raw_filename
+
+    return filename
+
+
+def download_file_from_ftp(ftp_instance: FTP,
+                           filename: str,
+                           output_filename: Path) -> None:
+    # check if already exists
+    if output_filename.exists():
+        print(f"File already exists! {output_filename}")
+        return
+
+    # download to output_filename
+    with open(output_filename, 'wb') as out_f:
+        ftp_instance.retrbinary("RETR " + filename, out_f.write)
+
+    if output_filename.exists():
+        print(f"Successful Download! {output_filename}")
+    else:
+        print(f"Error Downloading file: {output_filename}")
+
+    return
+
+
+def batch_ftp_request(args: Dict, filenames: List) -> None:
+    # unpack multiple arguments
+    raw_folder = args['raw_folder']
+    years = args['years']
+
+    legit_filenames = [f for f in filenames if str(years) in f]
+    # create one FTP connection for each batch
+    with FTP('ftp.star.nesdis.noaa.gov') as ftp:
+        ftp.login()
+        ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')
+
+        # download each filename using this FTP object
+        for raw_filename in legit_filenames:
+            output_filename = (
+                make_filename(raw_folder, raw_filename, dataset='vhi')
+            )
+            download_file_from_ftp(ftp, raw_filename, output_filename)
+
+    return
```

Comment:
These blank return statements aren't necessary

Reply:
They appear in some other functions too

Reply:
done thanks gabi!